Voeg meer http methods toe aan core/http-methods

TimvdLippe opened the pull request:
Hiermee maken we het expliciet dat ook de HTTP methods die enkel (metadata) informatie ophalen toegestaan zijn. Deze methoden maken geen wijzigingen of lezen resources uit, maar kunnen gebruikt worden om informatie over het endpoint zelf uit te lezen.

Fixes Geonovum/KP-APIs#633

sanderke commented:
De regel is geen white list van toegestane methods. Enkel onjuist gebruik van methods is niet toegestaan (lastig te testen met enkel openapi.json).

Als we toch overgaan naar een white list lijkt `MAY` meer gepast in toevoeging dan `SHOULD`.

TimvdLippe commented:
Op basis van ons gesprek heb ik de lijst nu optioneel gemaakt en non-exhaustive. Daarmee laten we het open, maar geven we ook expliciet de meest gebruikte methodes aan. Tevens heb ik wat informatie die in de "note" stond in het statement gezet, om zo de normatieve regels en de verduidelijking te scheiden van elkaar.

joostfarla commented on sections/designRules.md:
```diff
@@ -135,11 +135,13 @@ Although the REST architectural style does not impose a specific protocol, REST
       <dt>Rationale</dt>
       <dd>
          The HTTP specifications offer a set of standard methods, where every method is designed with explicit semantics. Adhering to the HTTP specification is crucial, since HTTP clients and middleware applications rely on standardized characteristics.
+         <p>According to <a href="https://www.rfc-editor.org/rfc/rfc9110#name-overview">RFC 9110 9.1</a> the <code>GET</code> and <code>HEAD</code> HTTP methods MUST be supported by the server, all other methods are optional.
```
Wat wordt hier bedoeld met `MUST be supported by the server`? Op welke resources? Bij operation resources kan het zijn dat er geen `GET` (en dus ook `HEAD`) operatie bestaat.

sanderke replied:
Hiermee wordt bedoeld dat de server het dient te ondersteunen. Of het wordt toegestaan wordt per resource bepaald. Zin toegevoegd om dit onderscheid te verduidelijken.

TimvdLippe replied:
Laten we deze zin weghalen en de suggestie van Alexander toevoegen aan de volgende paragraaf

TimvdLippe replied:
```suggestion
```

TimvdLippe commented:
Actiepunt van TO: @sanderke HTTP RFC opnieuw lezen en kijken wat de intentie is. Tevens inlezen op OpenAPI 3.2 waar meer methoden worden toegevoegd (https://github.com/OAI/OpenAPI-Specification/issues/1747)

sanderke commented on sections/designRules.md:
```diff
@@ -135,11 +135,13 @@ Although the REST architectural style does not impose a specific protocol, REST
       <dt>Statement</dt>
       <dd>
-         Resources MUST be retrieved or manipulated using standard HTTP methods (GET/POST/PUT/PATCH/DELETE).
+         Resources MUST be retrieved or manipulated using standard HTTP methods (GET/POST/PUT/PATCH/DELETE). Resources MAY support HTTP methods that provide information without manipulating the underlying resources (including but not limited to HEAD/CONNECT/OPTIONS/TRACE).
```
```suggestion
         An API MUST adhere to the HTTP method semantics defined in [rfc9110].
```

sanderke commented on sections/designRules.md:
```diff
@@ -174,54 +176,73 @@ Although the REST architectural style does not impose a specific protocol, REST
+         <div class="note">
+            <p>The HTTP specification [[rfc9110]] offers a set of standard methods, where every method is designed with explicit semantics.
+            The OpenAPI Specification 3.0 <a href="https://spec.openapis.org/oas/v3.0.1#path-item-object">Path Item Object</a> also supports these methods, except for <code>CONNECT</code>.
+            <p>In addition to the standard HTTP methods, a server may support other optional methods as well, e.g. <code>PROPFIND</code>, <code>COPY</code>, <code>PURGE</code>, <code>VIEW</code>, <code>LINK</code>, <code>UNLINK</code>, <code>LOCK</code>, <code>UNLOCK</code>, etc.
```
```suggestion
            <p>In addition to the standard HTTP methods, a server may support other optional methods as well, e.g. <code>PROPFIND</code>, <code>COPY</code>, <code>PURGE</code>, <code>VIEW</code>, <code>LINK</code>, <code>UNLINK</code>, <code>LOCK</code>, <code>UNLOCK</code>, etc. Additional methods MAY be documented using OpenAPI 3.2 while preserving these semantics defined in [rfc9110].
```

TimvdLippe replied:
Ik weet niet of we hier al kunnen verwijzen naar 3.2. Misschien kunnen we dit in een niet-normatieve noot toevoegen van "dit komt er in de toekomst aan"?

sanderke commented on sections/designRules.md:
```diff
@@ -135,11 +135,13 @@ Although the REST architectural style does not impose a specific protocol, REST
       <dd>
          The HTTP specifications offer a set of standard methods, where every method is designed with explicit semantics. Adhering to the HTTP specification is crucial, since HTTP clients and middleware applications rely on standardized characteristics.
+         <p>According to <a href="https://www.rfc-editor.org/rfc/rfc9110#name-overview">RFC 9110 9.1</a> the <code>GET</code> and <code>HEAD</code> HTTP methods MUST be supported by the server, all other methods are optional.
+         <p>If an optional HTTP request method is sent to a server and the server does not support that HTTP method for the target resource, an HTTP status code <code>405 Method Not Allowed</code> shall be returned and a list of allowed methods for the target resource shall be provided in the <code>Allow</code> header in the response as stated in <a href="https://www.rfc-editor.org/rfc/rfc9110#na ...
```
```suggestion
         <p>The set of methods allowed is determined per target resource. If an optional HTTP request method is sent to a server and the server does not support that HTTP method for the target resource, an HTTP status code <code>405 Method Not Allowed</code> shall be returned and a list of allowed methods for the target resource shall be provided in the <code>Allow</code> header in the response as stated in <a href="https://www.rfc-editor.org/rfc/rfc9110#name-405-method-not-allowed">RFC 9110 15.5.6</a>.
```

sanderke commented on sections/designRules.md:
```diff
@@ -174,54 +176,73 @@ Although the REST architectural style does not impose a specific protocol, REST
+         </div>
       </dd>
       <dt>How to test</dt>
       <dd>
-         Analyse the OpenAPI Description to confirm all supported methods are either `post`, `put`, `get`, `delete`, or `patch`.
+         Analyse the OpenAPI Description to confirm all supported methods are either `get`, `post`, `put`, `patch`, `delete`, `head`, `options` or `trace`.
       </dd>
```
Deze begrensde lijst zou niet meer opgaan.

TimvdLippe replied:
Aangezien het sowieso anders niet in een OpenAPI specificatie kan worden gezet, moeten we volgens mij deze hele zin weghalen en precies dat opschrijven.

TimvdLippe commented on sections/designRules.md:
```diff
@@ -130,16 +130,18 @@ A resource that corresponds to a single conceptual entity is referred to as a [=
       <dt>Statement</dt>
       <dd>
-         Resources MUST be retrieved or manipulated using standard HTTP methods (GET/POST/PUT/PATCH/DELETE).
+         An API MUST adhere to the HTTP method semantics defined in [rfc9110].
```
```suggestion
         An API MUST adhere to the HTTP method semantics defined in [[rfc9110]].
```

TimvdLippe commented:
Deze wijziging is goedgekeurd bij het TO 2025-12-02, maar dan enkel de wijziging in het statement. De wijzigingen in het rationale zijn niet goedgekeurd en vervallen.

TimvdLippe committed:
Maak semantiek van HTTP expliciet in /core/http-methods

Hiermee maken we het expliciet dat ook de HTTP methods
die enkel (metadata) informatie ophalen toegestaan zijn.
Deze methoden maken geen wijzigingen of lezen resources uit,
maar kunnen gebruikt worden om informatie over het endpoint
zelf uit te lezen.
